### PR TITLE
[4] remove space in output html class list

### PR DESCRIPTION
**Pull request by PhilETaylor**
### Summary of Changes

Removes a space in the default HTML output for the cassiopeia template

### Testing Instructions

Visit Joomla 4 home page after a default install with no sample data

View source 

### Actual result BEFORE applying this Pull Request

NOTE THE SPACE at the end of the classes "header container-header full-width"

Screenshot 2021-04-30 at 23 27 35

Also removes space here (before has-side-bar-right): 

```html
<body class="site-grid site com_content wrapper-static view-featured no-layout no-task itemid-101  has-sidebar-right">
```

### Expected result AFTER applying this Pull Request

Screenshot 2021-04-30 at 23 30 46

```html
<body class="site-grid site com_content wrapper-static view-featured no-layout no-task itemid-101 has-sidebar-right">
```

### CODE REVIEW

**Comment by brianteeman** [comment]
Sorry this is not correct as it results in $pageclass.$hasclass

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 30e1e168fd43182b2762f0df6de654547a69ad90<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33456">issues.joomla.org/tracker/joomla-cms/33456</a>.</sub>

**Comment by PhilETaylor** [comment]
ok try again now - improved it.

**Comment by brianteeman** [comment]
honestly as long as there is a space between classes who cares if its one or two

**Comment by PhilETaylor** [comment]
I care enough to fix it. It makes the output look amateurish. We can do better.

**Comment by brianteeman** [comment]
better to be maintainable and understandable

**Comment by PhilETaylor** [comment]
And on that note Im going to bed. There is nothing non-maintainable or non-understandable about this PR or trying to do better...

**Comment by PhilETaylor** [comment]
For example - this sucks as output! 

```html

</div>
	<div class="controls">
		<input
	type="email"
	inputmode="email"
	name="jform[email]"
	 class="form-control validate-email required"	id="jform_email"
	value=""
	 size="30"     autocomplete="email"    required  >			</div>
</div>

```

**Comment by Quy** [comment]
Here is one way and then `implode`.

https://github.com/joomla/joomla-cms/blob/cc70dbddd29ca586c87574f5a4486e60d28bee9b/layouts/joomla/form/field/time.php#L50

**Comment by PhilETaylor** [comment]
Imploding on `$attributes` actually GIVES double empty spaces :) unless you first filter out the blank array entries.

**Comment by PhilETaylor** [comment]
> @richard67 added the Updates Requested  label 6 hours ago

This PR is ready for testing, no further updates are needed, it produces maintainable and understandable output. 

The commentary was on other improvements around output that can be made to the email fields (Which is on my todo list)

**Comment by richard67** [comment]
@PhilETaylor I don't think that this is right: `$hasClass .= 'has-sidebar-left';`. I think the space which was there before was right.

**Comment by PhilETaylor** [comment]
Dont think - test :) you will see the final output code it correct. Line 117 correctly outputs the space that was removed from 84/89

**Comment by richard67** [comment]
> Dont think - test :) you will see the final output code it correct. Line 117 correctly outputs the space that was removed from 84/89

No need to test if reading the code shows it's wrong. If in line 84 the variable has a value, let's say 'gaga', you append 'has-sidebar-left' to it when there is a module on the left sidebar, which results in 'gagahas-sidebar-left', and that's wrong.

And if there was no value in line 84 so it ends in 'has-sidebar-left', you append later in line 'has-sidebar-right' to it when there is a module on the right sidebar, and the ends in 'has-sidebar-lefthas-sidebar-right'. How shall line 117 clean that up? I don't see how it can do that.

**Comment by PhilETaylor** [comment]
r_1475504_PEFNR

**Comment by PhilETaylor** [comment]
added back one space - its so ugly having strings that start with a space like " this" - just makes code so ugly

**Comment by brianteeman** [comment]
> just makes code so ugly

or makes it work

**Comment by PhilETaylor** [comment]
Thanks for your helpful insight once again - I wish I can be like you when I grow up.

File: templates/cassiopeia/index.php [hunk 1]
```diff
@@ -81,7 +81,7 @@
 
 if ($this->countModules('sidebar-left', true))
 {
-	$hasClass .= ' has-sidebar-left';
+	$hasClass .= 'has-sidebar-left';
```
Comment (richard67) [inline templates/cassiopeia/index.php]:
```suggestion
	$hasClass .= ' has-sidebar-left';
```

Comment (PhilETaylor) [reply templates/cassiopeia/index.php:84]: No. A space is prepended on Line 117

Comment (richard67) [reply templates/cassiopeia/index.php]: That one needs to be changed, back, too, then I'm ok.

Comment (PhilETaylor) [reply templates/cassiopeia/index.php]: Done.

**Comment by richard67** [comment]
Thanks @PhilETaylor .

File: templates/cassiopeia/index.php [hunk 2]
```diff
@@ -113,11 +113,11 @@
 	. ($layout ? ' layout-' . $layout : ' no-layout')
 	. ($task ? ' task-' . $task : ' no-task')
 	. ($itemid ? ' itemid-' . $itemid : '')
-	. ' ' . $pageclass
-	. $hasClass;
-	echo ($this->direction == 'rtl' ? ' rtl' : '');
+	. ($pageclass ? ' ' . $pageclass : '')
+	. ($hasClass ? ' ' . $hasClass : '')
```
Comment (Quy) [inline templates/cassiopeia/index.php]: This will produce double spaces.

Comment (PhilETaylor) [reply templates/cassiopeia/index.php]: Done.

File: templates/cassiopeia/index.php [hunk 3]
```diff
@@ -113,11 +113,11 @@
 	. ($layout ? ' layout-' . $layout : ' no-layout')
 	. ($task ? ' task-' . $task : ' no-task')
 	. ($itemid ? ' itemid-' . $itemid : '')
-	. ' ' . $pageclass
-	. $hasClass;
-	echo ($this->direction == 'rtl' ? ' rtl' : '');
+	. ($pageclass ? ' ' . $pageclass : '')
+	. ($hasClass ? $hasClass : '')
```
Comment (Quy) [inline templates/cassiopeia/index.php]:
```suggestion
	. $hasClass
```

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on acc7384bcb157247367dcb800980b9b3097c71f9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33456">issues.joomla.org/tracker/joomla-cms/33456</a>.</sub>

**Comment by ChristineWk** [comment]
I have tested this item :white_check_mark: successfully on acc7384bcb157247367dcb800980b9b3097c71f9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33456">issues.joomla.org/tracker/joomla-cms/33456</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33456">issues.joomla.org/tracker/joomla-cms/33456</a>.</sub>

**Comment by richard67** [comment]
Drone failure is not related to this PR, see issue #33477 .

**Comment by richard67** [comment]
I've updated the branch to latest 4.0-dev to get rid of the unrelated javascript-cs error in drone and restored the previous test results in the issue tracker.

**Comment by chmst** [comment]
Thanks!

**Pull request merged**